### PR TITLE
add  foreach support for custom device 

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes #ISSUE_NUMBER
for custom device, we want to support foreach, so I add a func that we could set other device type, and the default value is cuda.

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
I have a question about a design decision here—why should we have a default_device_type in SupportForeachDeviceType instead of just a list of devices that foreach works on? i.e., why not have SupportForeachDevices be a collection of devices that support foreach?

**Comment by heidongxianhua** [comment]
> I have a question about a design decision here—why should we have a default_device_type in SupportForeachDeviceType instead of just a list of devices that foreach works on? i.e., why not have SupportForeachDevices be a collection of devices that support foreach?

yeah,there are two points. Firstly, we want to reuse the default device type in optimize.py. And,now for other features, we use the similar mode to make it support custom device. And in the future,we will only give a similar func like SupportForeachDeviceType for all of features to make the default device is custom device replace cuda. So we just give a default device rather than a list of device type.@janeyx99

**Comment by heidongxianhua** [comment]
@janeyx99 could you have a look again？

**Comment by janeyx99** [comment]
Yes, sorry I am still not quite sure if I understand the two points all the way. Will leave more comments in the code itself.

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:10]: Why not
```suggestion
class SupportForeachDevices(object):
```

I understand the idea of storing devices that support foreach, but I _don't_ understand why we have a _default_device_type attribute that we replace.

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:15]:
```suggestion
    _default_devices = ["cuda"]
```
Why not have a list of supported devices? Will there never be the case where there may be tensors on multiple devices (across CUDA and the custom device type)? Even if there isn't, why is it better to specify only 1 supported device over a general list of supported devices?

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:50]:
```suggestion
    def get_foreach_supported_device_types_and(device: str) -> List[str]:
```
Is this the intention?

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:93]: Ah, if this is the only usage, maybe we should just add cpu to the SupportForeachDevices list?

File: torch/optim/adam.py
Comment (janeyx99) [inline torch/optim/adam.py:44]: So this is in particular is not just for foreach but the fused adam kernel, which requires cuda (torch.fused_adam is only defined for CUDA). The check here would not be appropriate.

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:29]: What's the use case for this?

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:15]: I appreciate the class wrapping a list, but what's the advantage of having a class instead of just having this list be explicitly by itself? Could you go into detail about how the use case with a custom device?

Comment (janeyx99) [inline torch/utils/_foreach_utils.py:15]: For example, this does feel more unsafe, but what do we benefit from the class wrapper?

```suggestion
__all__ = ["foreach_supported_devices"]

foreach_supported_devices = ["cuda"]
```

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:10]: Bikeshed on naming: ForeachSupportedDevices perhaps?
```suggestion
class ForeachSupportedDevices(object):
```

**Review by janeyx99 — COMMENTED** [review]
Hello! Thanks for the quick turnaround. I left a few questions, but I think I would like to verify the exact use case before approving. I'm curious how this new class will be used for custom devices. My perception is that the custom device code will add its name to the foreach supported device list and then the foreach ops should work from that extension point.

So then this leads to
1. Is my perception of the use case correct? Please correct me if not!
2. Do we need a class wrapper over just a simple accessible list? Do you plan on adding more to the class?

File: torch/optim/adam.py
Comment (heidongxianhua) [reply torch/optim/adam.py:44]: > particular

@janeyx99  yeah, we are cuda-like device, so we will support the foreach-op and fused adam/adamw.

**Comment by heidongxianhua** [comment]
> Hello! Thanks for the quick turnaround. I left a few questions, but I think I would like to verify the exact use case before approving. I'm curious how this new class will be used for custom devices. My perception is that the custom device code will add its name to the foreach supported device list and then the foreach ops should work from that extension point.
> 
> So then this leads to
> 
> 1. Is my perception of the use case correct? Please correct me if not!
> 2. Do we need a class wrapper over just a simple accessible list? Do you plan on adding more to the class?

yes,what you said is all right.  In fact, at the beginning, I wanted to directly modify these judgment device types and just add custom device. But I think this will make many hard code like "cuda","cpu" and "foo" for custom device,so I use a class named supportdefaultdevice to do this, and give more funcs to make it work well with fuesd and foreach in optimizer. But now I think it's may lead to more complex logic,  so I  think that modify the device judgement and add "foo" is ok. And waiting for your reply.@janeyx99

File: torch/optim/adam.py
Comment (janeyx99) [reply torch/optim/adam.py:44]: Right, but the fused_adam op has stricter requirements than the foreach and is fundamentally different. I would prefer not conflating the two. Instead, it may do well to expose a list of fused_kernel_supported devices in this file that people can add their custom device to.

**Comment by heidongxianhua** [comment]
> 

yes,what you said is all right.  In fact, at the beginning, I wanted to directly modify these judgment device types and just add custom device. But I think this will make many hard code like "cuda","cpu" and "foo" for custom device,so I use a class named supportdefaultdevice to do this, and give more funcs to make it work well with fuesd and foreach in optimizer. But now I think it's may lead to more complex logic,  so I  think that modify the device judgement and add "foo" is ok. And waiting for your reply.@janeyx99

> I think having an explicit list (instead of a class) should be sufficient. I agree it wouldn't be great to hardcode anything into the has_foreach_support function, but having an accessible list that it calls would be good enough abstraction.

enen, yeah, we just add two list may be good, foreach_kernels_supported_devices = ["cpu","cuda","foo"], fused_kernels_supported_devices=["cuda","foo"]. Waiting for your reply.@janeyx99

**Comment by janeyx99** [comment]
Yes, I like the new simpler plan. I just want to make sure it works well with your use case too. If it does, feel free to make the revisions and I will review!

**Comment by heidongxianhua** [comment]
> Yes, I like the new simpler plan. I just want to make sure it works well with your use case too. If it does, feel free to make the revisions and I will review!

yeah, I have made changes to add two list, and now the list of `foreach_kernels_supported_devices` only contains "cuda", "foo". It will be used in the func named `_default_to_fused_or_foreach` defined in `torch/optim/optimizer.py`, and when I add "cpu", the testcase will failed with `error dynamo/test_optimizers.py::OptimizerTests::test_adadelta `. I do not get the root cause, so I just add "cuda", "foo" to `foreach_kernels_supported_devices` to keep consistent with existing code logic. @janeyx99

**Comment by heidongxianhua** [comment]
@janeyx99  I have made some changes, maybe you could review again

File: torch/optim/adam.py
Comment (janeyx99) [inline torch/optim/adam.py:47]:
```suggestion
            fused_supported_devices = get_fused_kernels_supported_devices()
```

File: torch/optim/adam.py
Comment (janeyx99) [inline torch/optim/adam.py:46]: This isn't true. We group by device and dtype later on when calling torch.fused_adam so this requirement can be relaxed.

File: torch/optim/adamw.py
Comment (janeyx99) [inline torch/optim/adamw.py:1]: Same comments as in adam.py

File: torch/optim/adam.py
Comment (janeyx99) [inline torch/optim/adam.py:53]:
```suggestion
                raise RuntimeError("`fused=True` requires all the params to be floating point Tensors"
                                   f"of supported devices: {fused_supported_devices}")
```

File: torch/optim/optimizer.py
Comment (janeyx99) [inline torch/optim/optimizer.py:70]: the check_same_device shouldn't be necessary

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:53]: This should call get_foreach_kernels_supported_devices to avoid duplication

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:10]: Let's also keep these private by prefixing with underscore: _get_foreach_kernels_supported_devices and _get_fused_kernels_supported_devices

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:20]: Has torch._C._get_privateuse1_backend_name() already landed as how private backends should be referred to?

**Review by janeyx99 — COMMENTED** [review]
Thanks for modifying the code--it makes more sense and is more lightweight now :D 

Hmmm I'll check out the CI failure

File: torch/utils/_foreach_utils.py
Comment (heidongxianhua) [reply torch/utils/_foreach_utils.py:18]: > Has torch._C._get_privateuse1_backend_name() already landed as how private backends should be referred to?

yes, the API of torch._C._get_privateuse1_backend_name()  has been merged and used many times in pytorch.

**Comment by heidongxianhua** [comment]
> 

sorry, it is my mistake.

**Comment by heidongxianhua** [comment]
> Wait..is the CI failure still there?

sorry, it is my mistake. And I make some changes again according to your comment.

**Review by janeyx99 — COMMENTED** [review]
Looks good! CI failures look real, but a rebase may just fix as the failing tests seem to be disabled.

Once CI is green, feel free to ping me to approve

**Comment by heidongxianhua** [comment]
> Looks good! CI failures look real, but a rebase may just fix as the failing tests seem to be disabled.
> 
> Once CI is green, feel free to ping me to approve

yeah  I have rebase the pr and make all checks ok, maybe you could add approve label.@janeyx99

**Review by janeyx99 — APPROVED** [review]

**Comment by janeyx99** [comment]
@heidongxianhua sad, it looks like this PR broke inductor tests (see HUD link above). When relanding, we should make sure to add the label ciflow/inductor to verify green signal.

**Comment by heidongxianhua** [comment]
> @heidongxianhua sad, it looks like this PR broke inductor tests (see HUD link above). When relanding, we should make sure to add the label ciflow/inductor to verify green signal.

yeah, I make some tests, it seems like the func of ` torch._C._get_privateuse1_backend_name()` will cause graph break which break the ci workflow. And could you give me some suggestion to avoid graph break?
the func of ` torch._C._get_privateuse1_backend_name()` is used to get a string (the name of privateuse1 renamed by `torch._C._rename_privateuse1_backend()`) defined in C++.

**Comment by heidongxianhua** [comment]
> > @heidongxianhua sad, it looks like this PR broke inductor tests (see HUD link above). When relanding, we should make sure to add the label ciflow/inductor to verify green signal.
> 
> yeah, I make some tests, it seems like the func of ` torch._C._get_privateuse1_backend_name()` will cause graph break which break the ci workflow. And could you give me some suggestion to avoid graph break? the func of ` torch._C._get_privateuse1_backend_name()` is used to get a string (the name of privateuse1 backend renamed by `torch._C._rename_privateuse1_backend()`) defined in C++. waiting for your reply, thankyou. @janeyx99

@janeyx99  I have solved the inductor test failed, could you have a look again?

File: torch/utils/backend_registration.py
Comment (janeyx99) [inline torch/utils/backend_registration.py:90]: @albanD  for a second look here

**Comment by jansel** [comment]
> I'm not sure if the graph break is expected and whether there's a way to implement this without one. @jansel Would you have an idea or someone to redirect to?

@mlazos would be a good contact here.

File: torch/utils/backend_registration.py
Comment (albanD) [reply torch/utils/backend_registration.py:90]: Sounds ok to me.
Can we open an issue to track this @heidongxianhua please?
This will most likely involve teaching dynamo about `_get_privateuse1_backend_name` which shouldn't be too hard for someone that knows dynamo well.

Comment (heidongxianhua) [reply torch/utils/backend_registration.py:90]: > Sounds ok to me. Can we open an issue to track this @heidongxianhua please? This will most likely involve teaching dynamo about `_get_privateuse1_backend_name` which shouldn't be too hard for someone that knows dynamo well.

yeah, I will give an issue later. And could you approve it again to merge this? @janeyx99

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**
